File: legacy-python/cursor_orch/session.py
from __future__ import annotations

from pathlib import Path

import yaml

from cursor_orch.config import (
    OrchestratorConfig,
    RepoConfig,
    parse_config,
    to_yaml,
    validate_config,
)

SESSION_DIR = Path.home() / ".cursor-orch"
SESSION_PATH = SESSION_DIR / "session.yaml"
SETUP_STATE_PATH = SESSION_DIR / "setup-state.yaml"
VALID_SETUP_STEPS = {"model", "prompt", "confirm"}


class Session:
    def __init__(self) -> None:
        self._config = OrchestratorConfig(
            name="",
            model="",
        )
        self._setup_state: dict[str, str | bool] = {
            "active": False,
            "step": "model",
        }

    def set_name(self, name: str) -> None:
        self._config.name = name

    def set_model(self, model: str) -> None:
        self._config.model = model

    def set_prompt(self, prompt: str) -> None:
        self._config.prompt = prompt

    def add_repo(self, alias: str, url: str, ref: str = "main") -> bool:
        replaced = alias in self._config.repositories
        self._config.repositories[alias] = RepoConfig(url=url, ref=ref)
        return replaced

    def remove_repo(self, alias: str) -> bool:
        if alias in self._config.repositories:
            del self._config.repositories[alias]
            return True
        return False

    def set_branch_prefix(self, prefix: str) -> None:
        self._config.target.branch_prefix = prefix

    def set_auto_pr(self, enabled: bool) -> None:
        self._config.target.auto_create_pr = enabled

    def set_bootstrap_repo(self, name: str) -> None:
        self._config.bootstrap_repo_name = name

    @property
    def config(self) -> OrchestratorConfig:
        return self._config

    def build_config(self) -> OrchestratorConfig:
        return self._config

    def has_required_guided_values(self) -> bool:
        return bool(self._config.model.strip()) and bool(self._config.prompt.strip())

    def setup_state(self) -> dict[str, str | bool]:
        return dict(self._setup_state)

    def set_setup_state(self, *, active: bool | None = None, step: str | None = None) -> None:
        if active is not None:
            self._setup_state["active"] = active
        if step is not None:
            normalized_step = step if step in VALID_SETUP_STEPS else "model"
            self._setup_state["step"] = normalized_step

    def clear_setup_state(self) -> None:
        self._setup_state = {
            "active": False,
            "step": "model",
        }

    def should_resume_guided_setup(self) -> bool:
        active = bool(self._setup_state.get("active", False))
        if not active:
            return False
        step = str(self._setup_state.get("step", "model"))
        if step not in VALID_SETUP_STEPS:
            return True
        if step == "confirm":
            return True
        return not self.has_required_guided_values()

    def validate(self) -> list[str]:
        try:
            validate_config(self._config)
        except ValueError as exc:
            return [str(exc)]
        return []

    def save(self, path: Path) -> None:
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text(to_yaml(self._config))

    def load(self, path: Path) -> None:
        self._config = parse_config(path.read_text())

    def save_session(self) -> None:
        self.save(SESSION_PATH)
        self._save_setup_state()

    def load_session(self) -> bool:
        loaded = False
        if SESSION_PATH.exists():
            self.load(SESSION_PATH)
            loaded = True
        self._load_setup_state()
        return loaded

    def _save_setup_state(self) -> None:
        SESSION_DIR.mkdir(parents=True, exist_ok=True)
        payload = {
            "active": bool(self._setup_state.get("active", False)),
            "step": str(self._setup_state.get("step", "model")),
        }
        SETUP_STATE_PATH.write_text(yaml.safe_dump(payload, sort_keys=False))

    def _load_setup_state(self) -> None:
        if not SETUP_STATE_PATH.exists():
            self.clear_setup_state()
            return
        raw = yaml.safe_load(SETUP_STATE_PATH.read_text())
        if not isinstance(raw, dict):
            self.clear_setup_state()
            return
        active_raw = raw.get("active", False)
        step_raw = raw.get("step", "model")
        active = bool(active_raw)
        step = str(step_raw)
        if step not in VALID_SETUP_STEPS:
            step = "model"
        self._setup_state = {
            "active": active,
            "step": step,
        }

File: src/session.ts
import * as fs from "node:fs";
import { createHash } from "node:crypto";
import * as os from "node:os";
import * as path from "node:path";
import YAML from "yaml";
import type { McpServerConfig, OrchestratorConfig } from "./config/types.js";
import { parseConfig, parseMcpServers, toYaml } from "./config/parse.js";
import { validateConfig, validateMcpServers } from "./config/validate.js";

type SessionPaths = {
  sessionDir: string;
  sessionPath: string;
  setupStatePath: string;
  historyPath: string;
};

function hashSessionKey(sessionKey: string): string {
  return createHash("sha256").update(sessionKey).digest("hex").slice(0, 16);
}

function resolveSessionKey(cwd: string, sessionKey: string | undefined): string {
  if (sessionKey && sessionKey.trim()) {
    return sessionKey.trim();
  }
  return path.resolve(cwd);
}

export function resolveSessionPaths(opts?: {
  homeDir?: string;
  cwd?: string;
  sessionKey?: string;
}): SessionPaths {
  const homeDir = opts?.homeDir ?? os.homedir();
  const cwd = opts?.cwd ?? process.cwd();
  const sessionKey = resolveSessionKey(cwd, opts?.sessionKey ?? process.env.CURSOR_ORCH_SESSION_KEY);
  const sessionRoot = path.join(homeDir, ".cursor-orch");
  const sessionDir = path.join(sessionRoot, "sessions", hashSessionKey(sessionKey));
  return {
    sessionDir,
    sessionPath: path.join(sessionDir, "session.yaml"),
    setupStatePath: path.join(sessionDir, "setup-state.yaml"),
    historyPath: path.join(sessionDir, "history"),
  };
}

const defaultSessionPaths = resolveSessionPaths();
export const SESSION_DIR = defaultSessionPaths.sessionDir;
export const SESSION_PATH = defaultSessionPaths.sessionPath;
export const SETUP_STATE_PATH = defaultSessionPaths.setupStatePath;
export const VALID_SETUP_STEPS = new Set(["model", "prompt", "confirm"]);

function extractMcpServersPayload(raw: unknown): unknown {
  if (raw === null || raw === undefined) return {};
  if (typeof raw !== "object" || Array.isArray(raw)) {
    throw new Error("MCP servers file must contain a mapping");
  }
  const o = raw as Record<string, unknown>;
  if ("mcp_servers" in o && o.mcp_servers !== undefined) {
    return o.mcp_servers;
  }
  if ("mcpServers" in o && o.mcpServers !== undefined) {
    return o.mcpServers;
  }
  return o;
}

export function createDefaultOrchestratorConfig(): OrchestratorConfig {
  return {
    name: "",
    model: "",
    prompt: "",
    repositories: {},
    tasks: [],
    target: { auto_create_pr: true, consolidate_prs: true, branch_prefix: "cursor-orch", branch_layout: "consolidated" },
    bootstrap_repo_name: "cursor-orch-bootstrap",
    mcp_servers: {},
  };
}
const LEGACY_SESSION_ROOT = path.join(os.homedir(), ".cursor-orch");
const LEGACY_SESSION_PATH = path.join(LEGACY_SESSION_ROOT, "session.yaml");
const LEGACY_SETUP_STATE_PATH = path.join(LEGACY_SESSION_ROOT, "setup-state.yaml");

export class Session {
  private _config: OrchestratorConfig;
  private _setupState: { active: boolean; step: string };
  private readonly sessionPaths: SessionPaths;

  constructor(paths: SessionPaths = resolveSessionPaths()) {
    this.sessionPaths = paths;
    this._config = createDefaultOrchestratorConfig();
    this._setupState = { active: false, step: "model" };
  }

  setName(name: string): void {
    this._config.name = name;
  }

  setModel(model: string): void {
    this._config.model = model;
  }

  setPrompt(prompt: string): void {
    this._config.prompt = prompt;
  }

  addRepo(alias: string, url: string, ref = "main"): boolean {
    const replaced = alias in this._config.repositories;
    this._config.repositories[alias] = { url, ref };
    return replaced;
  }

  removeRepo(alias: string): boolean {
    if (alias in this._config.repositories) {
      delete this._config.repositories[alias];
      return true;
    }
    for (const [key, repo] of Object.entries(this._config.repositories)) {
      if (repo.url === alias) {
        delete this._config.repositories[key];
        return true;
      }
    }
    return false;
  }

  setBranchPrefix(prefix: string): void {
    this._config.target.branch_prefix = prefix;
  }

  setAutoPr(enabled: boolean): void {
    this._config.target.auto_create_pr = enabled;
  }

  setConsolidatePrs(enabled: boolean): void {
    this._config.target.consolidate_prs = enabled;
  }

  setBootstrapRepo(name: string): void {
    this._config.bootstrap_repo_name = name;
  }

  setMcpServer(name: string, server: McpServerConfig): void {
    validateMcpServers({ [name]: server });
    const current = this._config.mcp_servers ?? {};
    current[name] = server;
    this._config.mcp_servers = current;
  }

  removeMcpServer(name: string): boolean {
    const current = this._config.mcp_servers;
    if (current && name in current) {
      delete current[name];
      return true;
    }
    return false;
  }

  clearMcpServers(): void {
    this._config.mcp_servers = {};
  }

  importMcpServersFromMap(map: Record<string, McpServerConfig>): { added: string[]; replaced: string[] } {
    const parsed = parseMcpServers(map);
    validateMcpServers(parsed);
    const current = this._config.mcp_servers ?? {};
    const added: string[] = [];
    const replaced: string[] = [];
    for (const [name, server] of Object.entries(parsed)) {
      if (name in current) {
        replaced.push(name);
      } else {
        added.push(name);
      }
      current[name] = server;
    }
    this._config.mcp_servers = current;
    return { added, replaced };
  }

  importMcpServersFromFile(filePath: string): { added: string[]; replaced: string[] } {
    const content = fs.readFileSync(filePath, "utf8");
    const raw = YAML.parse(content);
    const candidate = extractMcpServersPayload(raw);
    return this.importMcpServersFromMap(candidate as Record<string, McpServerConfig>);
  }

  resetSessionToDefaults(): void {
    this._config = createDefaultOrchestratorConfig();
    this.clearSetupState();
  }

  get config(): OrchestratorConfig {
    return this._config;
  }

  get historyPath(): string {
    return this.sessionPaths.historyPath;
  }

  buildConfig(): OrchestratorConfig {
    return this._config;
  }

  hasRequiredGuidedValues(): boolean {
    return Boolean(this._config.model.trim()) && Boolean(this._config.prompt.trim());
  }

  setupState(): { active: boolean; step: string } {
    return { ...this._setupState };
  }

  setSetupState(opts: { active?: boolean; step?: string }): void {
    if (opts.active !== undefined) {
      this._setupState.active = opts.active;
    }
    if (opts.step !== undefined) {
      this._setupState.step = VALID_SETUP_STEPS.has(opts.step) ? opts.step : "model";
    }
  }

  clearSetupState(): void {
    this._setupState = { active: false, step: "model" };
  }

  shouldResumeGuidedSetup(): boolean {
    const active = this._setupState.active;
    if (!active) return false;
    const step = this._setupState.step;
    if (!VALID_SETUP_STEPS.has(step)) return true;
    if (step === "confirm") return true;
    return !this.hasRequiredGuidedValues();
  }

  validate(): string[] {
    try {
      validateConfig(this._config);
    } catch (e) {
      return [String(e)];
    }
    return [];
  }

  save(filePath: string): void {
    fs.mkdirSync(path.dirname(filePath), { recursive: true });
    fs.writeFileSync(filePath, toYaml(this._config), "utf8");
  }

  load(filePath: string): void {
    this._config = parseConfig(fs.readFileSync(filePath, "utf8"));
  }

  saveSession(): void {
    this.save(this.sessionPaths.sessionPath);
    this.saveSetupState();
  }

  loadSession(): boolean {
    const configPath = this.resolveReadablePath(this.sessionPaths.sessionPath, LEGACY_SESSION_PATH);
    const loaded = Boolean(configPath);
    if (configPath) {
      this.load(configPath);
    }
    this.loadSetupState();
    return loaded;
  }

  private saveSetupState(): void {
    fs.mkdirSync(this.sessionPaths.sessionDir, { recursive: true });
    const payload = {
      active: this._setupState.active,
      step: this._setupState.step,
    };
    fs.writeFileSync(this.sessionPaths.setupStatePath, YAML.stringify(payload), "utf8");
  }

  private loadSetupState(): void {
    const setupStatePath = this.resolveReadablePath(this.sessionPaths.setupStatePath, LEGACY_SETUP_STATE_PATH);
    if (!setupStatePath) {
      this.clearSetupState();
      return;
    }
    const raw = YAML.parse(fs.readFileSync(setupStatePath, "utf8"));
    if (typeof raw !== "object" || raw === null) {
      this.clearSetupState();
      return;
    }
    const o = raw as Record<string, unknown>;
    const active = Boolean(o.active);
    let step = String(o.step ?? "model");
    if (!VALID_SETUP_STEPS.has(step)) {
      step = "model";
    }
    this._setupState = { active, step };
  }

  private resolveReadablePath(primaryPath: string, fallbackPath: string): string | null {
    if (fs.existsSync(primaryPath)) {
      return primaryPath;
    }
    if (fs.existsSync(fallbackPath)) {
      return fallbackPath;
    }
    return null;
  }
}
